src/exchange.ts

```diff
@@ -0,0 +1,235 @@
+          function: "createOrder",
+          transaction: transferID,
+          pair: [pair.from, pair.to],
+          price: price,
+        },
+        [
+          {
+            name: "Exchange",
+            value: "Verto",
+          },
+          {
+            name: "Action",
+            value: "Order",
+          },
+          ...tags,
+        ]
+      );
+    }, [pair.from, this.utils.CLOB_CONTRACT, pair.to]);
+
+    if (orderID === null) throw new Error("Could not create order");
+
+    // Create exchange fee
+    await this.utils.createFee(amount, pair.from, orderID, "exchange");
+
+    // Create VRT holder fee
+    await this.utils.createFee(amount, pair.from, orderID, "token_holder");
+
+    // Call Discord hook
+    axios.post(`https://hook.verto.exchange/api/transaction?id=${orderID}`);
+
+    return orderID;
+  }
+
+  /**
+   * Cancel an order. This will return the non-filled tokens for an order and
+   * remove it from the orderbook
+   * @param orderID The transaction id of the swap
+   * @returns The transaction id of the cancel
+   */
+  async cancel(orderID: string): Promise<string> {
+    const contract = this.smartweave
+      .contract(this.utils.CLOB_CONTRACT)
+      .connect(this.wallet);
+
+    const transactionID = await cacheContractHook(
+      async () =>
+        contract.writeInteraction(
+          {
+            function: "cancelOrder",
+            orderID,
+          },
+          [
+            {
+              name: "Exchange",
+              value: "Verto",
+            },
+            {
+              name: "Action",
+              value: "CancelOrder",
+            },
+          ]
+        ),
+      this.utils.CLOB_CONTRACT
+    );
+
+    if (!transactionID) throw new Error("Order could not be cancelled");
+
+    return transactionID;
+  }
+
+  /**
+   * Fetches the order book for a specific token from the CLOB contract
+   * @param input Token contract ID or token pair
+   * @returns List of orders
+   */
+  async getOrderBook(input: string | TokenPair): Promise<OrderInterface[]> {
+    // get clob contract state
+    const clobContractState: {
+      [key: string]: any;
+      pairs: {
+        pair: [string, string];
+        orders: {
+          [key: string]: any;
+        }[];
+      }[];
+    } = await this.utils.getState(this.utils.CLOB_CONTRACT);
+
+    // map orders
+    const allOrders: OrderInterface[][] = clobContractState.pairs
+      .filter(({ pair }) => {
+        if (typeof input === "string") return pair.includes(input);
+        else return pair.includes(input[0]) && pair.includes(input[1]);
+      })
+      .map(({ pair, orders }) =>
+        orders.map((order) => ({
+          id: order.id,
+          owner: order.creator,
+          pair,
+          price: order.price,
+          filled: order.originalQuantity - order.quantity,
+          quantity: order.originalQuantity,
+        }))
+      );
+
+    // flatten orders
+    return ([] as OrderInterface[]).concat(...allOrders);
```

Comment:
This works but take a look at `array.flat()`

Reply:
True, could u commit that maybe? I'm not familiar with `arra.flat()`

Reply:
![image](https://user-images.githubusercontent.com/60560109/145089166-d74693fe-995e-43d3-a493-1552cae9a3ba.png)